In [47]:
import pandas as pd

In [48]:
df = pd.read_excel('3.xlsx')  

In [49]:
df

Соревнование  Год проведения  ID участника  Место команды участника Пол   
0       2014 весна            2014             1                      602   М  \
1       2014 весна            2014             2                       17   М   
2       2014 весна            2014             3                      122   М   
3       2014 весна            2014             4                      602   М   
4       2014 весна            2014             5                      344   М   
...            ...             ...           ...                      ...  ..   
16500   2018 осень            2018         16501                       35   М   
16501   2018 осень            2018         16502                      165   М   
16502   2018 осень            2018         16503                      165   Ж   
16503   2018 осень            2018         16504                      590   Ж   
16504   2018 осень            2018         16505                      590   Ж   

       Год рождения Город проживания  
0              1983           Москва  
1              1987           Москва  
2              1964      Новосибирск  
3              1982           Москва  
4              1983  Cанкт-Петербург  
...             ...              ...  
16500          1980           Москва  
16501          1981           Москва  
16502          1983           Москва  
16503          1978         Смоленск  
16504          1980         Смоленск  

[16505 rows x 7 columns]

In [50]:
df['Город проживания'] = df['Город проживания'].apply(lambda x: x.capitalize())

In [51]:
len(df['Город проживания'].unique())

363

In [52]:
def return_ok(string_defect):
    list_of_defects = ['Cанкт-петербург', 'Санкт-петербург', 'Ленинград-обнинск-москва', 'Ленинград', 'Спб', 'St. petersburg', 'Петербург']
    if string_defect in list_of_defects:
        return "Cанкт-петербург"
    else:
        return string_defect
    
df['Город проживания'] = df['Город проживания'].apply(lambda x: return_ok(x))            

Были выписаны все возможные варианты написания города 'Cанкт-петербург' и объеденены в 1, вариантов написания 'Екатеринбург' - 1, поэтому с ним не были проведены манипуляции. 

Учатников с городом "Ленинград-обнинск-москва" = 10, поэтому даже если их не считать, все равно "Cанкт-петербург" на первом месте

In [53]:
len(df['Город проживания'].unique())

357

In [54]:
df

Соревнование  Год проведения  ID участника  Место команды участника Пол   
0       2014 весна            2014             1                      602   М  \
1       2014 весна            2014             2                       17   М   
2       2014 весна            2014             3                      122   М   
3       2014 весна            2014             4                      602   М   
4       2014 весна            2014             5                      344   М   
...            ...             ...           ...                      ...  ..   
16500   2018 осень            2018         16501                       35   М   
16501   2018 осень            2018         16502                      165   М   
16502   2018 осень            2018         16503                      165   Ж   
16503   2018 осень            2018         16504                      590   Ж   
16504   2018 осень            2018         16505                      590   Ж   

       Год рождения Город проживания  
0              1983           Москва  
1              1987           Москва  
2              1964      Новосибирск  
3              1982           Москва  
4              1983  Cанкт-петербург  
...             ...              ...  
16500          1980           Москва  
16501          1981           Москва  
16502          1983           Москва  
16503          1978         Смоленск  
16504          1980         Смоленск  

[16505 rows x 7 columns]

In [76]:
def return_ok(string_defect):
    list_of_defects = ['Королёв', 'Королев']
    if string_defect in list_of_defects:
        return "Королев"
    else:
        return string_defect
    
df['Город проживания'] = df['Город проживания'].apply(lambda x: return_ok(x))  
print(len(df['Город проживания'].unique()))

356


In [77]:
def return_ok(string_defect):
    list_of_defects = ['Москва', 'Мск', 'Moscow', 'Москва ']
    if string_defect in list_of_defects:
        return "Москва"
    else:
        return string_defect
    
df['Город проживания'] = df['Город проживания'].apply(lambda x: return_ok(x))  
print(len(df['Город проживания'].unique()))

353


In [78]:
cur = df['Город проживания'].value_counts()

In [79]:
cur

Город проживания
Москва             9586
Cанкт-петербург    1508
Екатеринбург       1435
Зеленоград          170
Долгопрудный        159
                   ... 
Орехово-зеуво         1
Невьянск              1
Copenhagen            1
Череповец             1
Мо, лыткарино         1
Name: count, Length: 353, dtype: int64

In [86]:
cnt_cities = 0
for i in range(61):
    cnt_cities += cur[i]
print(cnt_cities / 16505)

0.930748258103605


In [89]:
for i in range(61):
    print(cur.index[i])

Москва
Cанкт-петербург
Екатеринбург
Зеленоград
Долгопрудный
Королев
Салехард
Рязань
Люберцы
Подольск
Мытищи
Химки
Одинцово
Железнодорожный
Красногорск
Балашиха
Воронеж
Тверь
Обнинск
Жуковский
Пушкино
Дубна
Тула
Электросталь
Лыткарино
Черноголовка
Раменское
Калуга
Лобня
Ногинск
Дмитров
Домодедово
Липецк
Фрязино
Пущино
Ивантеевка
Сергиев посад
Владимир
Тамбов
Лабытнанги
Ярославль
Шевлягино
Смоленск
Троицк
Краснознаменск
Таллин
Нижний новгород
Белгород
Коломна
Волгоград
Пос. ильинский
Щелково
Солнечногорск
Монино
Димитровград
Наро-фоминск
Реутов
Видное
Чехов
Электроугли
Дзержинский


Возьмем первые 61 город после чистки данных(они уникальны). Они покрывают 92 % от всех городов. Близжайший город к нашей 3, который может по итогу войти в топ 3 это "Зеленоград". Предположим, что все оставшиеся города(не всключая наши первые 61) неправильно написаны и каждый из учатников имел в виду "Зеленоград" => 170 + (1 - 0.93) * 16505 ≈ 1325, что меньше у наименьшего из топ 3(1435), а значит, что итоговая тройка не изменится.

### Жители города "Москва", "Cанкт-петербург", "Екатеринбург" больше всего представлены в соревнованиях